### PR TITLE
Review treatment of issued at, issuer, audience #16

### PR DESCRIPTION
**Pull request by dknoma**
Added a few more tests for testing with `iss` and `aud` claims, as well as changing existing tests to support the new feature (no `iss` or `aud` claims are treated as nulls).

### CODE REVIEW

**Review by jfallows — DISMISSED** [review]
Looks great!

We'll also want to add `resolveEx` to `OAuthFunctions` to layout the `issuer` and `audience`, but what you have right now is fine to get started.

**Comment by dknoma** [comment]
I successfully created another file, `TooManyRealmsKeys.jwk`, with the test realms specifically for the too many realms tests and all tests are passing. Changed the `fails.too.many.realms` controller and nukleus scripts to now use the correct kids `[realm1, realm2, ..., realm17TooMany]`.

**Comment by dknoma** [comment]
> Looks great!
> 
> We'll also want to add `resolveEx` to `OAuthFunctions` to layout the `issuer` and `audience`, but what you have right now is fine to get started.

So instead of using the regular `claims` method, we'll want `resolveEx` to add the `iss` and `aud` claims for us?

**Review by jfallows — DISMISSED** [review]
Once the dependencies are ready, this looks good to merge.

File: src/main/java/org/reaktivity/specification/nukleus/oauth/internal/OAuthJwtKeys.java [hunk 1]
```diff
@@ -82,70 +80,68 @@
...
+        {
+            rsKeyPair = initRSAKeyPair(rsa256);
+            esKeyPair = initECKeyPair(es256);
+        }
+        catch(Exception e)
+        {
+            // Utility
+        }
+        RFC7515_RS256 = rsKeyPair;
+        RFC7515_ES256 = esKeyPair;
```
Comment (jfallows) [inline src/main/java/org/reaktivity/specification/nukleus/oauth/internal/OAuthJwtKeys.java]: Recommend changing to simpler form:
```
        try
        {
            RFC7515_RS256 = initRSAKeyPair(rsa256);
            RFC7515_ES256 = initECKeyPair(es256);
        }
        catch(Exception ex)
        {
            throw new IllegalStateException(ex);
        }
```

File: src/main/java/org/reaktivity/specification/nukleus/oauth/internal/OAuthJwtKeys.java [hunk 2]
```diff
@@ -82,70 +80,68 @@
...
     }
 
     private static KeyPair initRSAKeyPair(
-        Map<String, String> params)
+        Map<String, String> params) throws Exception
```
Comment (jfallows) [inline src/main/java/org/reaktivity/specification/nukleus/oauth/internal/OAuthJwtKeys.java]: Make sure the `throws` exception type is as specific as possible, not just generic `Exception`.
Please apply this feedback to all such `throws` in the method signatures called from this method too, same for `initECKeyPair`.

File: src/main/scripts/org/reaktivity/specification/nukleus/oauth/control/resolve/one.realm.with.issuer.and.audience/controller.rpt [hunk 3]
```diff
@@ -31,8 +31,10 @@ write ${controlR05C.nextCorrelationId()}
 write ${core:string("oauth")}
 write [0x05] "RS256"
 write 0
-write [0x0B] "test issuer"      # issuer extension
-write [0x0D] "test audience"    # audience extension
+write ${oauth:resolveEx()
+              .issuer("test issuer")
+              .audience("test audience")
+              .build()}
```
Comment (jfallows) [inline src/main/scripts/org/reaktivity/specification/nukleus/oauth/control/resolve/one.realm.with.issuer.and.audience/controller.rpt]: Each `.` should be vertically aligned with the `:` for readability.
Please apply that approach to all scripts.

File: src/test/java/org/reaktivity/specification/nukleus/oauth/internal/OAuthJwtKeysTest.java [hunk 4]
```diff
@@ -24,10 +24,15 @@
...
 
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.ExpectedException;
 
 public final class OAuthJwtKeysTest
 {
+    @Rule
+    public final ExpectedException thrown = ExpectedException.none();
+
```
Comment (jfallows) [inline src/test/java/org/reaktivity/specification/nukleus/oauth/internal/OAuthJwtKeysTest.java]: Needed?

File: src/main/java/org/reaktivity/specification/nukleus/oauth/internal/OAuthFunctions.java [hunk 5]
```diff
@@ -68,6 +77,23 @@ public static JwtHelper jwt(
...
 
+    static JwtHelper jwt(
+        KeyPair pair,
+        String kind,
+        String algorithm)
+    {
+        Map<String, Supplier<JwtHelper>> helperFactories = new HashMap<>();
+        helperFactories.put(kind, () -> new JwtHelper(pair, kind, algorithm));
+        Supplier<JwtHelper> helperFactory = helperFactories.get(kind);
+        return helperFactory.get();
```
Comment (jfallows) [inline src/main/java/org/reaktivity/specification/nukleus/oauth/internal/OAuthFunctions.java]: No need to populate a `helperFactories` map, just need to `return new JwtHelper(pair, kind, algorithm)` directly.

**Review by jfallows — CHANGES_REQUESTED** [review]

File: src/main/java/org/reaktivity/specification/nukleus/oauth/internal/OAuthJwtKeys.java [hunk 6]
```diff
@@ -80,23 +83,19 @@
...
-            // Utility
+            throw new IllegalStateException(ex);
         }
-        RFC7515_RS256 = rsKeyPair;
-        RFC7515_ES256 = esKeyPair;
     }
 
     private static KeyPair initRSAKeyPair(
-        Map<String, String> params) throws Exception
+        Map<String, String> params) throws NoSuchAlgorithmException, InvalidKeySpecException
```
Comment (jfallows) [inline src/main/java/org/reaktivity/specification/nukleus/oauth/internal/OAuthJwtKeys.java]: Simplify to `GeneralSecurityException` which is a common super type for these exceptions.
Please apply to all similar method signatures.

File: src/main/java/org/reaktivity/specification/nukleus/oauth/internal/OAuthJwtKeys.java [hunk 7]
```diff
@@ -80,23 +83,19 @@
...
-        KeyPair esKeyPair = null;
         try
         {
-            rsKeyPair = initRSAKeyPair(rsa256);
-            esKeyPair = initECKeyPair(es256);
+            RFC7515_RS256 = initRSAKeyPair(rsa256);
+            RFC7515_ES256 = initECKeyPair(es256);
         }
-        catch(Exception e)
+        catch(Exception ex)
```
Comment (jfallows) [inline src/main/java/org/reaktivity/specification/nukleus/oauth/internal/OAuthJwtKeys.java]: Change catch to `GeneralSecurityException` if it is sufficient to cover all the cases.

File: src/test/java/org/reaktivity/specification/nukleus/oauth/internal/OAuthFunctionsTest.java [hunk 8]
```diff
@@ -81,20 +71,16 @@ public void shouldSignJWTwithES256() throws GeneralSecurityException
         assertNotNull(token);
     }
 
-    @Test
+    @Test(expected = RuntimeException.class)
```
Comment (jfallows) [inline src/test/java/org/reaktivity/specification/nukleus/oauth/internal/OAuthFunctionsTest.java]: Why is this not getting a `GeneralSecurityException` instead?

**Review by jfallows — CHANGES_REQUESTED** [review]

**Review by jfallows — APPROVED** [review]

**Pull request merged**